swagger_spec_validator/validator20.py

```diff
@@ -21,5 +22,101 @@ def validate_spec_url(spec_url):
 
 
 def validate_spec(spec_json):
+    """Validates a Swagger 2.0 API Specification given a json dict.
+
+    :param spec_json: the json dict of the swagger spec.
+    :returns: `None` in case of success, otherwise raises an exception.
+    :raises: :py:class:`swagger_spec_validator.SwaggerValidationError`
+    :raises: :py:class:`jsonschema.exceptions.ValidationError`
+    """
     validate_json(spec_json, 'schemas/v2.0/schema.json')
-    # TODO: rich schema validation
+
+    apis, definitions = spec_json['paths'], spec_json['definitions']
+
+    validate_apis(apis)
+
+    validate_definitions(definitions)
+
+
+def validate_apis(apis):
+    """Validates the semantic errors in `paths` of the Spec.
+
+    :param apis: dict of all the paths
+    :returns: `None` in case of success, otherwise raises an exception.
+    :raises: :py:class:`swagger_spec_validator.SwaggerValidationError`
+    :raises: :py:class:`jsonschema.exceptions.ValidationError`
+    """
+    for api_name in apis:
+        api_body = apis[api_name]
+        api_params = api_body.get('parameters', [])
+        validate_duplicate_param(api_params)
+        for oper_name in api_body:
+            oper_body = api_body[oper_name]
+            op_params = oper_body.get('parameters', [])
+            validate_duplicate_param(op_params)
```

Comment:
style: consistency for terms in the same method ```op_*``` vs ```oper_*```.  Pick one and stick to it.